src/Everscale/Address.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright © 2017-2021 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include <charconv>
+#include <system_error>
+
+#include "Address.h"
+#include "HexCoding.h"
+#include "Wallet.h"
+#include "WorkchainType.h"
+
+using namespace TW;
+
+namespace TW::Everscale {
+
+bool Address::isValid(const std::string& string) noexcept {
+    auto parsed = parseWorkchainId(string);
+    if (!parsed.has_value())
+        return false;
+
+    auto [workchainId, pos] = *parsed;
+
+    if (workchainId != WorkchainType::Basechain && workchainId != WorkchainType::Masterchain) {
+        return false;
+    }
+
+    if (string.size() != pos + 64) {
+        return false;
+    }
+
+    std::string addr = string.substr(pos);
+    return parse_hex(addr).size() == size;
+}
+
+Address::Address(const std::string& string) {
+    if (!Address::isValid(string)) {
+        throw std::invalid_argument("Invalid address string!");
+    }
+
+    auto parsed = parseWorkchainId(string);
+    auto [wc, pos] = *parsed;
+
+    _wc = wc;
+
+    const auto acc = parse_hex(string.substr(pos));
+    std::copy(begin(acc), end(acc), begin(_address));
+}
+
+Address::Address(const PublicKey& publicKey, int8_t workchainId) {
+    InitData initData(publicKey);
+    auto [wc, address] = initData.computeAddr(workchainId);
+
+    _wc = wc;
+    _address = address;
+}
+
+std::string Address::string() const {
+    std::string string = std::to_string(_wc) + ":" + hex(_address);
```

Comment:
(minor) please change variable name for something less confusing (e.g. result) :)